indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedResourceManagementStrategy.java

```diff
@@ -0,0 +1,360 @@
+        log.info("%s provisioning. Current wait time: %s", currentlyProvisioning, durSinceLastProvision);
+        if (durSinceLastProvision.isLongerThan(config.getMaxScalingDuration().toStandardDuration())) {
+          log.makeAlert("Worker node provisioning taking too long!")
+             .addData("millisSinceLastProvision", durSinceLastProvision.getMillis())
+             .addData("provisioningCount", currentlyProvisioning.size())
+             .emit();
+
+          workerConfig.getAutoScaler().terminateWithIds(Lists.newArrayList(currentlyProvisioning));
+          currentlyProvisioning.clear();
+        }
+      }
+
+      return didProvision;
+    }
+  }
+
+  int getScaleUpNodeCount(
+      final RemoteTaskRunnerConfig remoteTaskRunnerConfig,
+      final WorkerBehaviorConfig workerConfig,
+      final Collection<Task> pendingTasks,
+      final Collection<ZkWorker> zkWorkers
+  )
+  {
+    final int minWorkerCount = workerConfig.getAutoScaler().getMinNumWorkers();
+    final int maxWorkerCount = workerConfig.getAutoScaler().getMaxNumWorkers();
+    final Predicate<ZkWorker> isValidWorker = ResourceManagementUtil.createValidWorkerPredicate(config);
+    final int currValidWorkers = Collections2.filter(zkWorkers, isValidWorker).size();
+    int moreWorkersNeeded = currValidWorkers == 0 ? minWorkerCount : getWorkersNeededToAssignTasks(
+        remoteTaskRunnerConfig,
+        workerConfig,
+        pendingTasks,
+        zkWorkers
+    );
+
+    int want = Math.max(
+        minWorkerCount - currValidWorkers,
+        // Additional workers needed to reach minWorkerCount
+        Math.min(config.getMaxScalingStep(), moreWorkersNeeded)
+        // Additional workers needed to run current pending tasks
+    );
+
+    if (want > 0 && currValidWorkers >= maxWorkerCount) {
+      log.warn("Unable to provision more workers. Current workerCount[%d] maximum workerCount[%d].");
+      return 0;
+    }
+    want = Math.min(want, maxWorkerCount - currValidWorkers);
+    return want;
+  }
+
+  int getWorkersNeededToAssignTasks(
+      final RemoteTaskRunnerConfig remoteTaskRunnerConfig,
+      final WorkerBehaviorConfig workerConfig,
+      final Collection<Task> pendingTasks,
+      final Collection<ZkWorker> zkWorkers
+  )
+  {
+    final Collection<ZkWorker> validWorkers = Collections2.filter(
+        zkWorkers,
+        ResourceManagementUtil.createValidWorkerPredicate(config)
+    );
+
+    Map<String, ImmutableZkWorker> workersMap = Maps.newHashMap();
+    for (ZkWorker worker : validWorkers) {
+      workersMap.put(worker.getWorker().getHost(), worker.toImmutable());
+    }
+    WorkerSelectStrategy workerSelectStrategy = workerConfig.getSelectStrategy();
+    int need = 0;
+    int capacity = getAverageWorkerCapacity(zkWorkers);
+    for (Task task : pendingTasks) {
+      Optional<ImmutableZkWorker> selectedWorker = workerSelectStrategy.findWorkerForTask(
+          remoteTaskRunnerConfig,
+          ImmutableMap.copyOf(workersMap),
```

Comment:
the interface for WorkerSelectStrategy explicitly needs ImmutableMap.I think its done in order to prevent any races due to addition or removal of workers.